src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -110,21 +111,35 @@ public static void DeleteFile(string fullPath)
         public static FileAttributes GetAttributes(string fullPath)
         {
             Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA data = default;
-            int errorCode = FillAttributeInfo(fullPath, ref data, returnErrorOnNotFound: true);
-            if (errorCode != 0)
-                throw Win32Marshal.GetExceptionForWin32Error(errorCode, fullPath);
+            int errorCode = FillAttributeInfo(fullPath, ref data, true);
+            return errorCode != 0
+                ? throw Win32Marshal.GetExceptionForWin32Error(errorCode, fullPath)
+                : (FileAttributes)data.dwFileAttributes;
+        }
 
-            return (FileAttributes)data.dwFileAttributes;
+        public static FileAttributes GetAttributes(SafeFileHandle fileHandle)
+        {
+            Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA data = default;
+            int errorCode = FillAttributeInfo(fileHandle, ref data);
+            return errorCode != 0
+                ? throw Win32Marshal.GetExceptionForWin32Error(errorCode, fileHandle.Path)
+                : (FileAttributes)data.dwFileAttributes;
```

Comment:
Rather than duplicating all these methods, can we just have a shared helper?
```C#
internal static FileAttributes GetAttributes(string? fullPath, SafeFileHandle? fileHandle)
{
    Debug.Assert((fullPath is not null) ^ (fileHandle is not null));
    Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA data = default;
    int errorCode = fullPath is not null ? FillAttributeInfo(fullPath, ref data) : FillAttributeInfo(fileHandle, ref data);
    return errorCode != 0
        ? throw Win32Marshal.GetExceptionForWin32Error(errorCode, fileHandle.Path)
        : (FileAttributes)data.dwFileAttributes;
}
```
?

Reply:
@stephentoub 
On the one hand, is it a common practice to include two parameters, which one is decided afterwards, which one is not null?

On the other hand, the method signature of FillAttributeInfo is different for the SafeFileHandle overload, since a FileHandle cannot be non-found.

Also, for the `SafeFileHandle` overload we use `fileHandle.Path` instead of `fullPath`.

In my opinion the following suggestion would be as 2 helper methods. Would you agree with this?

```cs
        internal static Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA GetAttributeData(string? fullPath)
        {
            Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA data = default;
            int errorCode = FillAttributeInfo(fullPath, ref data, false);
            return errorCode != 0
                ? throw Win32Marshal.GetExceptionForWin32Error(errorCode, fullPath)
                : data;
        }

        internal static Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA GetAttributeData(SafeFileHandle fileHandle)
        {
            Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA data = default;
            int errorCode = FillAttributeInfo(fileHandle, ref data);
            return errorCode != 0
                ? throw Win32Marshal.GetExceptionForWin32Error(errorCode, fileHandle.Path)
                : data;
        }
```

Implementation would look like this:
```cs
        public static FileAttributes GetAttributes(string fullPath) => GetAttributeData(fullPath).dwFileAttributes;

        public static FileAttributes GetAttributes(SafeFileHandle fileHandle) => GetAttributeData(fileHandle).dwFileAttributes;
```

Reply:
See commit https://github.com/dotnet/runtime/pull/60507/commits/985347ecdccd2784287477bdf7300e97cac2189d